quickemu.1
.\" Automatically generated by Pandoc 2.2.3.2
.\"
.TH "QUICKEMU" "1" "January 7, 2022" "quickemu" "Quickemu User Manual"
.hy
.SH NAME
.PP
quickemu \- A quick VM builder and manager
.SH SYNOPSIS
.PP
\f[B]quickemu\f[] [\f[I]OPTION\f[]]\&...
.SH DESCRIPTION
.PP
\f[B]quickemu\f[] will create and run highly optimised desktop virtual
machines for Linux, macOS and Windows
.SH OPTIONS
.TP
.B \f[B]\[en]vm\f[]
vm configuration file
.RS
.RE
.PP
You can also pass optional parameters
.TP
.B \f[B]\[en]delete\f[]
Delete the disk image.
.RS
.RE
.TP
.B \f[B]\[en]display\f[]
Select display backend.
`sdl' (default), `gtk', `none' or `spice'
.RS
.RE
.TP
.B \f[B]\[en]fullscreen\f[]
Starts VM in full screen mode (Ctl+Alt+f to exit)
.RS
.RE
.TP
.B \f[B]\[en]ignore\-msrs\-always\f[]
Configure KVM to always ignore unhandled machine\-specific registers
.RS
.RE
.TP
.B \f[B]\[en]screen <screen>\f[]
Use specified screen to determine the window size.
.RS
.RE
.TP
.B \f[B]\[en]shortcut\f[]
Create a desktop shortcut
.RS
.RE
.TP
.B \f[B]\[en]snapshot apply <tag>\f[]
Apply/restore a snapshot.
.RS
.RE
.TP
.B \f[B]\[en]snapshot create <tag>\f[]
Create a snapshot.
.RS
.RE
.TP
.B \f[B]\[en]snapshot delete <tag>\f[]
Delete a snapshot.
.RS
.RE
.TP
.B \f[B]\[en]snapshot info\f[]
Show disk/snapshot info.
.RS
.RE
.TP
.B \f[B]\[en]status\-quo\f[]
Do not commit any changes to disk/snapshot.
.RS
.RE
.TP
.B \f[B]\[en]version\f[]
Print version
.RS
.RE
.SH EXAMPLES
.TP
.B \f[B]quickemu \[en]vm ubuntu\-mate\-impish.conf\f[]
Launches the VM specified in the file \f[I]ubuntu\-mate\-impish.conf\f[]
.RS
.RE
.SS Introduction
.PP
Quickly create and run highly optimised desktop virtual machines for
Linux, macOS and Windows; with just two commands.
You decide what operating system you want to run and Quickemu will
figure out the best way to do it for you.
For example:
.IP
.nf
\f[C]
quickget\ ubuntu\-mate\ impish
quickemu\ \-\-vm\ ubuntu\-mate\-impish.conf
\f[]
.fi
.PP
The original objective of the project was to enable quick testing of
Linux distributions where the virtual machine configurations can be
stored anywhere, such as external USB storage or your home directory,
and no elevated permissions are required to run the virtual machines.
\f[B]Quickemu now also includes comprehensive support for macOS and
Windows\f[].
.SS Features
.IP \[bu] 2
\f[B]macOS\f[] Monterey, Big Sur, Catalina, Mojave & High Sierra
.IP \[bu] 2
\f[B]Windows\f[] 8.1, 10 and 11 including TPM 2.0
.IP \[bu] 2
Ubuntu (https://ubuntu.com/desktop) and all the \f[B]official Ubuntu
flavours (https://ubuntu.com/download/flavours)\f[]
.IP \[bu] 2
Debian (https://www.debian.org/) (bullseye with all the official and
non\-free DE variants)
.IP \[bu] 2
Linux Mint (https://linuxmint.com/) ( Cinnamon , MATE , and XFCE) ,
elementary OS (https://elementary.io/) ,
Pop!_OS (https://pop.system76.com/) , ZorinOS (https://zorin.com/os/)
.IP \[bu] 2
Fedora (https://getfedora.org/) , Alma Linux (https://almalinux.org/) ,
Oracle Linux (https://www.oracle.com/linux/) and Rocky
Linux (https://rockylinux.org/)
.IP \[bu] 2
openSUSE (https://get.opensuse.org/) (
Leap (https://get.opensuse.org/leap/) ,
Tumbleweed (https://get.opensuse.org/tumbleweed/) ,
MicroOS (https://microos.opensuse.org/) )
.IP \[bu] 2
Arch Linux (https://www.archlinux.org/) , Kali (https://www.kali.org/) ,
Garuda (https://garudalinux.org/) & NixOS (https://nixos.org/)
.IP \[bu] 2
Regolith Linux (https://regolith-linux.org/) ( Release 1.6 and latest
2.0.0 pre\-release )
.IP \[bu] 2
FreeBSD (https://www.freebsd.org/) & OpenBSD (https://www.openbsd.org/)
.IP \[bu] 2
Full SPICE support including host/guest clipboard sharing
.IP \[bu] 2
VirtIO\-webdavd file sharing for Linux and Windows guests
.IP \[bu] 2
VirtIO\-9p file sharing for Linux and macOS guests
.IP \[bu] 2
QEMU Guest Agent support (https://wiki.qemu.org/Features/GuestAgent);
provides access to a system\-level agent via standard QMP commands
.IP \[bu] 2
Samba file sharing for Linux, macOS and Windows guests (\f[I]if
\f[CI]smbd\f[I] is installed on the host\f[])
.IP \[bu] 2
VirGL acceleration
.IP \[bu] 2
USB device pass\-through
.IP \[bu] 2
Smartcard pass\-through
.IP \[bu] 2
Automatic SSH port forwarding to guests
.IP \[bu] 2
Network port forwarding
.IP \[bu] 2
Full duplex audio
.IP \[bu] 2
EFI (with or without SecureBoot) and Legacy BIOS boot
.IP \[bu] 2
Graphical user interfaces available
.PP
Quickemu is a wrapper for the excellent QEMU (https://www.qemu.org/)
that attempts to automatically \f[I]\[lq]do the right thing\[rq]\f[],
rather than expose exhaustive configuration options.
.PP
We have a Discord for this project:
[IMAGE: Discord (https://img.shields.io/discord/712850672223125565?color=0C306A&label=WimpysWorld%20Discord&logo=Discord&logoColor=ffffff&style=flat-square)] (https://discord.gg/sNmz3uw)
.PP
See this (old) video where I explain some of my motivations for creating
Quickemu.
.PP
[IMAGE: Replace VirtualBox with Bash &
QEMU (https://img.youtube.com/vi/AOTYWEgw0hI/0.jpg)] (https://www.youtube.com/watch?v=AOTYWEgw0hI)
.SS Requirements
.IP \[bu] 2
QEMU (https://www.qemu.org/) (\f[I]6.0.0 or newer\f[]) \f[B]with GTK,
SDL, SPICE & VirtFS support\f[]
.IP \[bu] 2
bash (https://www.gnu.org/software/bash/) (\f[I]4.0 or newer\f[])
.IP \[bu] 2
Coreutils (https://www.gnu.org/software/coreutils/)
.IP \[bu] 2
EDK II (https://github.com/tianocore/edk2)
.IP \[bu] 2
grep (https://www.gnu.org/software/grep/)
.IP \[bu] 2
jq (https://stedolan.github.io/jq/)
.IP \[bu] 2
LSB (https://wiki.linuxfoundation.org/lsb/start)
.IP \[bu] 2
procps (https://gitlab.com/procps-ng/procps)
.IP \[bu] 2
python3 (https://www.python.org/)
.IP \[bu] 2
macrecovery (https://github.com/acidanthera/OpenCorePkg/tree/master/Utilities/macrecovery)
.IP \[bu] 2
mkisofs (http://cdrtools.sourceforge.net/private/cdrecord.html)
.IP \[bu] 2
usbutils (https://github.com/gregkh/usbutils)
.IP \[bu] 2
util\-linux (https://github.com/karelzak/util-linux)
.IP \[bu] 2
sed (https://www.gnu.org/software/sed/)
.IP \[bu] 2
spicy (https://gitlab.freedesktop.org/spice/spice-gtk)
.IP \[bu] 2
swtpm (https://github.com/stefanberger/swtpm)
.IP \[bu] 2
Wget (https://www.gnu.org/software/wget/)
.IP \[bu] 2
xdg\-user\-dirs (https://www.freedesktop.org/wiki/Software/xdg-user-dirs/)
.IP \[bu] 2
xrandr (https://gitlab.freedesktop.org/xorg/app/xrandr)
.IP \[bu] 2
zsync (http://zsync.moria.org.uk/)
.SS Ubuntu Guest
.PP
\f[C]quickget\f[] will automatically download an Ubuntu release and
create the virtual machine configuration.
.IP
.nf
\f[C]
quickget\ ubuntu\ focal
quickemu\ \-\-vm\ ubuntu\-focal.conf
\f[]
.fi
.IP \[bu] 2
Complete the installation as normal.
.IP \[bu] 2
Post\-install:
.RS 2
.IP \[bu] 2
Install the SPICE agent (\f[C]spice\-vdagent\f[]) to enable copy/paste
and USB redirection
.RS 2
.IP \[bu] 2
\f[C]sudo\ apt\ install\ spice\-vdagent\f[]
.RE
.IP \[bu] 2
Install the SPICE WebDAV agent (\f[C]spice\-webdavd\f[]) to enable file
sharing.
.RS 2
.IP \[bu] 2
\f[C]sudo\ apt\ install\ spice\-webdavd\f[]
.RE
.RE
.SS Ubuntu devel (daily\-live) images
.PP
\f[C]quickget\f[] can also download/refresh devel images via
\f[C]zsync\f[] for Ubuntu developers and testers.
.IP
.nf
\f[C]
quickget\ ubuntu\ devel
quickemu\ \-\-vm\ ubuntu\-devel.conf
\f[]
.fi
.PP
You can run \f[C]quickget\ ubuntu\ devel\f[] to refresh your daily
development image as often as you like, it will even automatically
switch to a new series.
.SS Ubuntu Flavours
.PP
All the official Ubuntu flavours are supported, just replace
\f[C]ubuntu\f[] with your preferred flavour.
.IP \[bu] 2
\f[C]kubuntu\f[]
.IP \[bu] 2
\f[C]lubuntu\f[]
.IP \[bu] 2
\f[C]ubuntu\f[]
.IP \[bu] 2
\f[C]ubuntu\-budgie\f[]
.IP \[bu] 2
\f[C]ubuntu\-kylin\f[]
.IP \[bu] 2
\f[C]ubuntu\-mate\f[]
.IP \[bu] 2
\f[C]ubuntu\-studio\f[]
.IP \[bu] 2
\f[C]xubuntu\f[]
.SS Other Linux Guests
.PP
\f[C]quickget\f[] also supports:
.IP \[bu] 2
\f[C]alma\f[]
.IP \[bu] 2
\f[C]archlinux\f[]
.IP \[bu] 2
\f[C]debian\f[]
.IP \[bu] 2
\f[C]elementary\f[]
.IP \[bu] 2
\f[C]fedora\f[]
.IP \[bu] 2
\f[C]garuda\f[]
.IP \[bu] 2
\f[C]kali\f[]
.IP \[bu] 2
\f[C]kdeneon\f[]
.IP \[bu] 2
\f[C]linuxmint\-cinnamon\f[]
.IP \[bu] 2
\f[C]linuxmint\-mate\f[]
.IP \[bu] 2
\f[C]linuxmint\-xfce\f[]
.IP \[bu] 2
\f[C]nixos\-gnome\f[]
.IP \[bu] 2
\f[C]nixos\-minimal\f[]
.IP \[bu] 2
\f[C]nixos\-plasma5\f[]
.IP \[bu] 2
\f[C]opensuse\f[]
.IP \[bu] 2
\f[C]oraclelinux\f[]
.IP \[bu] 2
\f[C]popos\f[]
.IP \[bu] 2
\f[C]regolith\f[]
.IP \[bu] 2
\f[C]rockylinux\f[]
.IP \[bu] 2
\f[C]solus\f[]
.IP \[bu] 2
\f[C]zorin\f[]
.PP
Or you can download a Linux image and manually create a VM
configuration.
.IP \[bu] 2
Download a .iso image of a Linux distribution
.IP \[bu] 2
Create a VM configuration file; for example
\f[C]debian\-bullseye.conf\f[]
.IP
.nf
\f[C]
guest_os="linux"
disk_img="debian\-bullseye/disk.qcow2"
iso="debian\-bullseye/firmware\-11.0.0\-amd64\-DVD\-1.iso"
\f[]
.fi
.IP \[bu] 2
Use \f[C]quickemu\f[] to start the virtual machine:
.IP
.nf
\f[C]
quickemu\ \-\-vm\ debian\-bullseye.conf
\f[]
.fi
.IP \[bu] 2
Complete the installation as normal.
.IP \[bu] 2
Post\-install:
.RS 2
.IP \[bu] 2
Install the SPICE agent (\f[C]spice\-vdagent\f[]) to enable copy/paste
and USB redirection.
.IP \[bu] 2
Install the SPICE WebDAV agent (\f[C]spice\-webdavd\f[]) to enable file
sharing.
.RE
.SS macOS Guest
.PP
\f[C]quickget\f[] automatically downloads a macOS recovery image and
creates a virtual machine configuration.
.IP
.nf
\f[C]
quickget\ macos\ catalina
quickemu\ \-\-vm\ macos\-catalina.conf
\f[]
.fi
.PP
macOS \f[C]high\-sierra\f[], \f[C]mojave\f[], \f[C]catalina\f[],
\f[C]big\-sur\f[] and \f[C]monterey\f[] are supported.
.IP \[bu] 2
Use cursor keys and enter key to select the \f[B]macOS Base System\f[]
.IP \[bu] 2
From \f[B]macOS Utilities\f[]
.RS 2
.IP \[bu] 2
Click \f[B]Disk Utility\f[] and \f[B]Continue\f[]
.RS 2
.IP \[bu] 2
On macOS Catalina, Big Sur & Monterey
.RS 2
.IP \[bu] 2
Select \f[C]Apple\ Inc.\ VirtIO\ Block\ Media\f[] from the list and
click \f[B]Erase\f[].
.RE
.IP \[bu] 2
On macOS Mojave and High Sierra
.RS 2
.IP \[bu] 2
Select \f[C]QEMU\ HARDDISK\ Media\f[] (~103.08GB) from the list and
click \f[B]Erase\f[].
.RE
.RE
.IP \[bu] 2
Enter a \f[C]Name:\f[] for the disk and click \f[B]Erase\f[].
.IP \[bu] 2
Click \f[B]Done\f[].
.IP \[bu] 2
Close Disk Utility
.RE
.IP \[bu] 2
From \f[B]macOS Utilities\f[]
.RS 2
.IP \[bu] 2
Click \f[B]Reinstall macOS\f[] and \f[B]Continue\f[]
.RE
.IP \[bu] 2
Complete the installation as you normally would.
.RS 2
.IP \[bu] 2
On the first reboot use cursor keys and enter key to select \f[B]macOS
Installer\f[]
.IP \[bu] 2
On the subsequent reboots use cursor keys and enter key to select the
disk you named
.RE
.PP
The default macOS configuration looks like this:
.IP
.nf
\f[C]
guest_os="macos"
img="macos\-catalina/RecoveryImage.img"
disk_img="macos\-catalina/disk.qcow2"
macos_release="catalina"
\f[]
.fi
.IP \[bu] 2
\f[C]guest_os="macos"\f[] instructs Quickemu to optimise for macOS.
.IP \[bu] 2
\f[C]macos_release="catalina"\f[] instructs Quickemu to optimise for a
particular macOS release.
.RS 2
.IP \[bu] 2
For example VirtIO Network and Memory Ballooning are available in Big
Sur and newer, but not previous releases.
.IP \[bu] 2
And VirtIO Block Media (disks) are supported/stable in Catalina and
newer.
.RE
.SS macOS compatibility
.PP
There are some considerations when running macOS via Quickemu.
.IP \[bu] 2
Supported macOS releases:
.RS 2
.IP \[bu] 2
High Sierra
.IP \[bu] 2
Mojave
.IP \[bu] 2
Catalina \f[B](Recommended)\f[]
.IP \[bu] 2
Big Sur
.IP \[bu] 2
Monterey
.RE
.IP \[bu] 2
\f[C]quickemu\f[] will automatically download the required
OpenCore (https://github.com/acidanthera/OpenCorePkg) bootloader and
OVMF firmware from OSX\-KVM (https://github.com/kholia/OSX-KVM).
.IP \[bu] 2
Optimised by default, but no GPU acceleration is available.
.RS 2
.IP \[bu] 2
Host CPU vendor is detected and guest CPU is optimised accordingly.
.IP \[bu] 2
VirtIO Block
Media (https://www.kraxel.org/blog/2019/06/macos-qemu-guest/) is used
for the system disk where supported.
.IP \[bu] 2
VirtIO \f[C]usb\-tablet\f[] (http://philjordan.eu/osx-virt/) is used for
the mouse.
.IP \[bu] 2
VirtIO Network (\f[C]virtio\-net\f[]) is supported and enabled on macOS
Big Sur and newer but previous releases use \f[C]vmxnet3\f[].
.IP \[bu] 2
VirtIO Memory Ballooning is supported and enabled on macOS Big Sur and
newer but disabled for other support macOS releases.
.RE
.IP \[bu] 2
USB host and SPICE pass\-through is:
.RS 2
.IP \[bu] 2
UHCI (USB 2.0) on macOS Catalina and earlier.
.IP \[bu] 2
XHCI (USB 3.0) on macOS Big Sur and newer.
.RE
.IP \[bu] 2
Display resolution can only be changed via macOS System Preferences.
.IP \[bu] 2
Full Duplex audio works on macOS High Sierra, Mojave and Catalina.
.RS 2
.IP \[bu] 2
\f[B]macOS Big Sur and Monterey have no audio at all\f[].
.RE
.IP \[bu] 2
File sharing between guest and host is available via
virtio\-9p (https://wiki.qemu.org/Documentation/9psetup) and SPICE
webdavd (https://gitlab.gnome.org/GNOME/phodav/-/merge_requests/24).
.IP \[bu] 2
Copy/paste via SPICE agent is \f[B]not available on macOS\f[].
.SS Windows 8.1, 10 & 11 Guests
.PP
\f[C]quickget\f[] can automatically download Windows 8.1, Windows
10 (https://www.microsoft.com/en-gb/software-download/windows10ISO) and
Windows 11 (https://www.microsoft.com/en-gb/software-download/windows11)
along with the VirtIO drivers for
Windows (https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/)
and creates a virtual machine configuration.
.IP
.nf
\f[C]
quickget\ windows\ 11
quickemu\ \-\-vm\ windows\-11.conf
\f[]
.fi
.IP \[bu] 2
Complete the installation as you normally would.
.IP \[bu] 2
All relevant drivers and services should be installed automatically.
.SS Regional versions
.PP
By default \f[C]quickget\f[] will download the \f[I]\[lq]English
International\[rq]\f[] release, but you can optionally specify one of
the supported languages: For example:
.IP
.nf
\f[C]
quickget\ windows\ 11\ "Chinese\ (Traditional)"
\f[]
.fi
.PP
The default Windows 11 configuration looks like this:
.IP
.nf
\f[C]
guest_os="windows"
disk_img="windows\-11/disk.qcow2"
iso="windows\-11/Win11_EnglishInternational_x64.iso"
fixed_iso="windows\-11/virtio\-win.iso"
tpm="on"
\f[]
.fi
.IP \[bu] 2
\f[C]guest_os="windows"\f[] instructs \f[C]quickemu\f[] to optimise for
Windows.
.IP \[bu] 2
\f[C]fixed_iso=\f[] specifies the ISO image that provides VirtIO
drivers.
.IP \[bu] 2
\f[C]tpm="on"\f[] instructs \f[C]quickemu\f[] to create a software
emulated TPM device using \f[C]swtpm\f[].
.SS BSD Guests
.PP
\f[C]quickemu\f[] supports FreeBSD and OpenBSD.
.IP
.nf
\f[C]
quickget\ freebsd\ 13_0
quickemu\ \-\-vm\ freebsd\-13_0.conf
\f[]
.fi
.SS Other Guests
.IP \[bu] 2
\f[C]android\f[]
.SH All the options
.PP
Here are the usage instructions:
.IP
.nf
\f[C]


Usage
\ \ quickemu\ \-\-vm\ ubuntu.conf

You\ can\ also\ pass\ optional\ parameters
\ \ \-\-delete\-disk\ \ \ \ \ \ \ \ \ \ \ :\ Delete\ the\ disk\ image\ and\ EFI\ variables
\ \ \-\-delete\-vm\ \ \ \ \ \ \ \ \ \ \ \ \ :\ Delete\ the\ entire\ VM\ and\ it\[aq]s\ configuration
\ \ \-\-display\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ :\ Select\ display\ backend.\ \[aq]sdl\[aq]\ (default),\ \[aq]gtk\[aq],\ \[aq]none\[aq],\ or\ \[aq]spice\[aq]
\ \ \-\-fullscreen\ \ \ \ \ \ \ \ \ \ \ \ :\ Starts\ VM\ in\ full\ screen\ mode\ (Ctl+Alt+f\ to\ exit)
\ \ \-\-ignore\-msrs\-always\ \ \ \ :\ Configure\ KVM\ to\ always\ ignore\ unhandled\ machine\-specific\ registers
\ \ \-\-screen\ <screen>\ \ \ \ \ \ \ :\ Use\ specified\ screen\ to\ determine\ the\ window\ size.
\ \ \-\-shortcut\ \ \ \ \ \ \ \ \ \ \ \ \ \ :\ Create\ a\ desktop\ shortcut
\ \ \-\-snapshot\ apply\ <tag>\ \ :\ Apply/restore\ a\ snapshot.
\ \ \-\-snapshot\ create\ <tag>\ :\ Create\ a\ snapshot.
\ \ \-\-snapshot\ delete\ <tag>\ :\ Delete\ a\ snapshot.
\ \ \-\-snapshot\ info\ \ \ \ \ \ \ \ \ :\ Show\ disk/snapshot\ info.
\ \ \-\-status\-quo\ \ \ \ \ \ \ \ \ \ \ \ :\ Do\ not\ commit\ any\ changes\ to\ disk/snapshot.
\ \ \-\-version\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ :\ Print\ version
\f[]
.fi
.SS Desktop shortcuts
.PP
Desktop shortcuts can be created for a VM, the shortcuts are saved in
\f[C]~/.local/share/applications\f[].
Here is an example of how to create a shortcut.
.IP
.nf
\f[C]
quickemu\ \-\-vm\ ubuntu\-focal\-desktop.conf\ \-\-shortcut
\f[]
.fi
.SS Screen and window size (Linux guests only)
.PP
\f[C]qemu\f[] will always default to the primary monitor to display the
VM's window.
.PP
Without the \f[C]\-\-screen\f[] option, \f[C]quickemu\f[] will look for
the size of the smallest monitor, and use a size that fits on said
monitor.
.PP
The \f[C]\-\-screen\f[] option forces \f[C]quickemu\f[] to use the size
of the given monitor to compute the size of the window.
\f[B]It won't use that monitor to display the VM's window if it's not
the primary monitor\f[].
This is useful if the primary monitor if not the smallest one, and if
the VM's window doesn't need to be moved around.
.PP
The \f[C]\-\-screen\f[] option is also useful with the
\f[C]\-\-fullscreen\f[] option, again because \f[C]qemu\f[] will always
use the primary monitor.
In order for the fullscreen mode to work properly, the resolution of the
VM's window must match the resolution of the screen.
.PP
To know which screen to use, type:
.IP
.nf
\f[C]
xrandr\ \-\-listmonitors\ |\ grep\ \-v\ Monitors
\f[]
.fi
.PP
The command will output something like this:
.IP
.nf
\f[C]
\ 0:\ +*HDMI\-0\ 2560/597x1440/336+1920+0\ \ HDMI\-0
\ 1:\ +DVI\-D\-0\ 1920/527x1080/296+0+0\ \ DVI\-D\-0
\f[]
.fi
.PP
The first number is what needs to be passed to the \f[C]\-\-screen\f[]
option.
.PP
For example:
.IP
.nf
\f[C]
quickemu\ \-\-vm\ vm.conf\ \-\-screen\ 0
\f[]
.fi
.PP
The above uses the 2560x1440 screen to compute the size of the window,
which Quickemu sizes to 2048x1152.
Without the \f[C]\-\-screen\f[] option, Quickemu would have used the
1920x1080 monitor which results in a window size of 1664x936.
.SH References
.PP
Useful reference that assisted the development of Quickemu.
.IP \[bu] 2
General
.RS 2
.IP \[bu] 2
QEMU's documentation! (https://qemu.readthedocs.io/en/latest/)
.IP \[bu] 2
<https://pve.proxmox.com/wiki/Qemu/KVM_Virtual_Machines>
.IP \[bu] 2
<https://www.kraxel.org/blog/2020/01/qemu-sound-audiodev/>
.RE
.IP \[bu] 2
macOS
.RS 2
.IP \[bu] 2
<https://www.nicksherlock.com/2020/06/installing-macos-big-sur-on-proxmox/>
.IP \[bu] 2
<https://passthroughpo.st/mac-os-adds-early-support-for-virtio-qemu/>
.IP \[bu] 2
<https://github.com/kholia/OSX-KVM>
.IP \[bu] 2
<https://github.com/thenickdude/KVM-Opencore>
.IP \[bu] 2
<https://github.com/acidanthera/OpenCorePkg/tree/master/Utilities/macrecovery>
.IP \[bu] 2
<https://www.kraxel.org/blog/2017/09/running-macos-as-guest-in-kvm/>
.IP \[bu] 2

quickemu.sh
#!/usr/bin/env bash

function vm_delete() {
  if [ -f "${disk_img}" ]; then
    rm "${disk_img}"
    echo "SUCCESS! Deleted ${disk_img}"
  fi
  exit 0
}

function vm_restore() {
  if [ -f "${disk_img_snapshot}" ]; then
    mv "${disk_img_snapshot}" "${disk_img}"
  fi
  echo "SUCCESS! Restored ${disk_img_snapshot}"
  exit 0
}

function vm_snapshot() {
  if [ -f "${disk_img_snapshot}" ]; then
    mv "${disk_img_snapshot}" "${disk_img_snapshot}.old"
  fi
  qemu-img create -b "${disk_img}" -f qcow2 "${disk_img_snapshot}" -q
  if [ $? -eq 0 ]; then
    echo "SUCCESS! Created ${disk_img_snapshot}"
  else
    echo "ERROR! Failed to create ${disk_img_snapshot}"
  fi
  exit 0
}

function get_port() {
    local PORT_START=22220
    local PORT_RANGE=9
    while true; do
        local CANDIDATE=$[${PORT_START} + (${RANDOM} % ${PORT_RANGE})]
        (echo "" >/dev/tcp/127.0.0.1/${CANDIDATE}) >/dev/null 2>&1
        if [ $? -ne 0 ]; then
            echo "${CANDIDATE}"
            break
        fi
    done
}

function vm_boot() {
  local VMNAME=$(basename ${VM} .conf)
  local VMDIR=$(dirname ${disk_img})
  local BIOS=""
  local GL="on"
  local VIRGL="on"
  local UI="sdl"
  local QEMU_VER=$(${QEMU} -version | head -n1 | cut -d' ' -f4 | cut -d'(' -f1)
  echo "Starting ${VM}"
  echo " - QEMU:     ${QEMU} v${QEMU_VER}"

  if [ -n "${disk_img}" ]; then
    disk_img_snapshot="${disk_img}.snapshot"
  else
    echo "ERROR! No disk_img defined."
    exit 1
  fi  

  readonly disk_min_size=395264
  if [ -z "${disk}" ]; then
    disk="64G"
  fi

  if [ ${ENABLE_EFI} -eq 1 ]; then
    if [ -e /snap/qemu-virgil/current/usr/share/qemu/edk2-x86_64-code.fd ] ; then
      if [ ! -e ${VMDIR}/${VMNAME}-vars.fd ]; then
        cp /snap/qemu-virgil/current/usr/share/qemu/edk2-i386-vars.fd ${VMDIR}/${VMNAME}-vars.fd
      fi
      BIOS="-drive if=pflash,format=raw,readonly,file=/snap/qemu-virgil/current/usr/share/qemu/edk2-x86_64-code.fd -drive if=pflash,format=raw,file=${VMDIR}/${VMNAME}-vars.fd"
    else
      echo " - EFI:      Booting requested but no EFI firmware found."
      echo "             Booting from Legacy BIOS."
    fi
    echo " - BIOS:     EFI"
  else
    echo " - BIOS:     Legacy"
  fi

  echo " - Disk:     ${disk_img} (${disk})"
  # If the disk is present but doesn't appear to have an install, then
  # remove it.
  if [ -e ${disk_img} ]; then
    local disk_curr_size=$(stat -c%s "${disk_img}")
    if [ ${disk_curr_size} -le ${disk_min_size} ]; then
      echo "             Looks unused, recreating."
      rm "${disk_img}"
    fi
  fi

  if [ ! -f "${disk_img}" ]; then
    # If there is no disk image, create a new image.
    ${QEMU_IMG} create -q -f qcow2 "${disk_img}" "${disk}"
    if [ $? -ne 0 ]; then
      echo "ERROR! Failed to create ${disk_img} of ${disk}. Stopping here."
      exit 1
    fi
    echo " - ISO:      ${iso}"
  else
    # If there is a disk image, do not boot from the iso
    iso=""
  fi
  if [ -e ${disk_img_snapshot} ]; then
    echo " - Snapshot: ${disk_img_snapshot}"
  fi

  local cores="1"
  local allcores=$(nproc --all)
  if [ ${allcores} -ge 8 ]; then
    cores="4"
  elif [ ${allcores} -ge 4 ]; then
    cores="2"
  fi
  echo " - CPU:      ${cores} Core(s)"

  local ram="2G"
  local allram=$(free --mega -h | grep Mem | cut -d':' -f2 | cut -d'G' -f1 | sed 's/ //g')
  if [ ${allram} -ge 64 ]; then
    ram="4G"
  elif [ ${allram} -ge 16 ]; then
    ram="3G"
  fi
  echo " - RAM:      ${ram}"

  # Determine what display to use
  local display="-display ${UI},gl=${GL}"
  echo " - UI:       ${UI}"
  echo " - GL:       ${GL}"
  echo " - VIRGL:    ${VIRGL}"

  local xres=1152
  local yres=648
  if [ "${XDG_SESSION_TYPE}" == "x11" ]; then
    local LOWEST_WIDTH=$(xrandr --listmonitors | grep -v Monitors | cut -d' ' -f4 | cut -d'/' -f1 | sort | head -n1)
    if [ ${LOWEST_WIDTH} -ge 3840 ]; then
      xres=3200
      yres=1800
    elif [ ${LOWEST_WIDTH} -ge 2560 ]; then
      xres=2048
      yres=1152
    elif [ ${LOWEST_WIDTH} -ge 1920 ]; then
      xres=1664
      yres=936
    elif [ ${LOWEST_WIDTH} -ge 1280 ]; then
      xres=1152
      yres=648
    fi
  fi
  echo " - Display:  ${xres}x${yres}"


  local NET=""
  # If smbd is available, export $HOME to the guest via samba
  if [ -e /snap/qemu-virgil/current/usr/sbin/smbd ]; then
      NET=",smb=${HOME}"
  fi

  if [ -n "${NET}" ]; then
    echo " - smbd:     ${HOME} will be exported to the guest via smb://10.0.2.4/qemu"
  else
    echo " - smbd:     ${HOME} will not be exported to the guest. 'smbd' not found."
  fi

  # Find a free port to expose ssh to the guest
  local PORT=$(get_port)
  if [ -n "${PORT}" ]; then
    NET="${NET},hostfwd=tcp::${PORT}-:22"
    echo " - ssh:      ${PORT}/tcp is connected. Login via 'ssh user@localhost -p ${PORT}'"
  else
    echo " - ssh:      All ports for exposing ssh have been exhausted."
  fi

  # Boot the iso image
  ${QEMU} -name ${VMNAME},process=${VMNAME} \
    ${BIOS} \
    -cdrom "${iso}" \
    -drive "file=${disk_img},format=qcow2,if=virtio,aio=native,cache.direct=on" \
    -enable-kvm \
    -machine q35,accel=kvm \
    -cpu host,kvm=on \
    -m ${ram} \
    -smp ${cores} \
    -net nic,model=virtio \
    -net user"${NET}" \
    -rtc base=localtime,clock=host \
    -serial mon:stdio \
    -audiodev pa,id=pa,server=unix:$XDG_RUNTIME_DIR/pulse/native,out.stream-name=${LAUNCHER}-${VMNAME},in.stream-name=${LAUNCHER}-${VMNAME} \
    -device intel-hda -device hda-duplex,audiodev=pa \
    -usb -device usb-kbd -device usb-tablet \
    -object rng-random,id=rng0,filename=/dev/urandom \
    -device virtio-rng-pci,rng=rng0 \
    -device virtio-vga,virgl=${VIRGL},xres=${xres},yres=${yres} \
    ${display} \
    "$@"
}

function usage() {
  echo
  echo "Usage"
  echo "  ${LAUNCHER} --vm ubuntu.conf"
  echo
  echo "You can also pass optional parameters"
  echo "  --delete   : Delete the disk image."
  echo "  --efi      : Enable EFI BIOS."
  echo "  --restore  : Restore the snapshot."
  echo "  --snapshot : Create a disk snapshot."
  exit 1
}

DELETE=0
ENABLE_EFI=0
readonly QEMU="/snap/bin/qemu-virgil"
readonly QEMU_IMG="/snap/bin/qemu-virgil.qemu-img"
readonly LAUNCHER=$(basename $0)
RESTORE=0
SNAPSHOT=0
VM=""

while [ $# -gt 0 ]; do
  case "${1}" in
    -efi|--efi)
      ENABLE_EFI=1
      shift;;
    -delete|--delete)
      DELETE=1
      shift;;
    -restore|--restore)
      RESTORE=1
      shift;;
    -snapshot|--snapshot)
      SNAPSHOT=1
      shift;;
    -vm|--vm)
      VM="$2"
      shift
      shift;;
    -h|--h|-help|--help)
      usage;;
    *)
      echo "ERROR! \"${1}\" is not a supported parameter."
      usage;;
  esac
done

# Check we have qemu-virgil available
if [ ! -e "${QEMU}" ] && [ ! -e "${QEMU_IMG}" ]; then
  echo "ERROR! qemu-virgil not found. Please install the qemu-virgil snap."
  echo "       https://snapcraft.io/qemu-virgil"
  exit 1
fi

if [ -n "${VM}" ] || [ -e "${VM}" ]; then
  source "${VM}"
else
  echo "ERROR! Virtual machine configuration not found."
  usage
fi

if [ ${DELETE} -eq 1 ]; then
  vm_delete
fi

if [ ${RESTORE} -eq 1 ]; then
  vm_restore
fi

if [ ${SNAPSHOT} -eq 1 ]; then
  vm_snapshot
fi

vm_boot